### ISSUE TITLE
aesmd daemon does not work

### ISSUE THREAD

**Issue by insujang** [01org/linux-sgx#69]
I am using both master branches for intel-sgx (SDK/PSW) and intel-sgx-driver.
Until yesterday, sgx sample applications worked well, however, not I cannot run all of them.

After investigating some reason, I found the following message in dmesg:
```
init: aesmd main process (...), terminated with status 1
init: aesmd main process ended, respawning
...
init: aesmd respawning too fast, stopped
```

And then I saw /var/log/syslog, which says to be that:
```
aesm_service[1346]: SGX Service unavailable
aesm_service[1346]: Fail to start service.
kernel: init: aesmd main process (1346) terminated with status 1
kernel: init: aesmd main process ended, respawning
...
```

Hence, creating enclave was failed when I tried to run samples due to lack of aesmd.
I also saw that this is known issue of SGX SDK 1.7 release.
How can I solve this problem?

Thanks.

**Comment by insujang** [comment]
I recompiled with DEBUG=1 flag and reinstalled it, then it works. 
I don't think this flag made it work, because service failed 10 times before working with syslog message:
```
aesm_service[16725]: The server sock is 0x2563c20
aesm_service[1281]: The server sock is 0x1e6d370
```

Was it problem of Intel's server? Then does it mean that aesm service is connected all the times for guaranteeing enclave's integrity even we don't use remote attestation?

**Comment by llly** [comment]
Uninstall psw and sgx-driver then install them. If not work, check the sgx flag in BIOS.

**Comment by insujang** [comment]
Well, it does not work, and problem becomes even worse.
I think the main problem of Intel's signature. Here is some tests that I did.

1. Compiling with `USE_OPT_LIBS=0`
README says this flag does not use intel's prebuilt enclave binaries that have been signed by Intel. With this flag, aesm service always shows "terminated with status 1".
2. Compiling without `USE_OPT_LIBS=0`
This case would make SGX work 2 weeks ago. However, now, my KVM kernel module shows "KVM internal error. Suberror: 1" and a VM freezes while installing psw.

Following is the environment that I tested.
1. Host
Intel Core i7 6700: SGX enabled by BIOS
Linux 14.04.5 LTS 64bit with kernel version 4.6.0+ (cloned from [kvm-sgx](https://github.com/01org/kvm-sgx))
QEMU-SGX (cloned from [qemu-sgx](https://github.com/01org/qemu-sgx))
Launching a qemu-kvm quest machine with `-epc 32M`, `-cpu host`, and `-enable-kvm`.

2. Guest
CPU: host (Intel Core i7 6700)
Linux 14.04.5 LT 64bit with kernel bersion 4.4.0.62

VM freezes right after printing the following command:
```
insujang@ubuntu:~/sgx/linux-sgx/linux/installer/bin$ sudo ./sgx_linux_x64_psw_1.7.100.36470.bin 
Unpacking Intel SGX PSW Package ... done.
Verifying the integrity of the install package ... done.
Installing Intel SGX PSW Package ... done.
/tmp/sgx-psw-zmcKU2 ~/sgx/linux-sgx/linux/installer/bin
install -d /opt/intel/sgxpsw
install -d /opt/intel/sgxpsw/scripts
install -d /usr/lib
cp -r package/* /opt/intel/sgxpsw
install scripts/* /opt/intel/sgxpsw/scripts
mv /opt/intel/sgxpsw/lib64/libsgx_uae_service.so /usr/lib
mv /opt/intel/sgxpsw/lib64/libsgx_urts.so /usr/lib
rmdir /opt/intel/sgxpsw/lib64
~/sgx/linux-sgx/linux/installer/bin
Installing aesmd.conf service ... done.
```

KVM prints the following kernel fault message:
```
KVM internal error. Suberror: 1
emulation failure
RAX=000000000000000a RBX=0000000000000003 RCX=ffffc90004001000 RDX=0000000000000000
RSI=ffff88000123fcc8 RDI=ffffc90004001000 RBP=ffff880273fcfc48 RSP=ffff880273fcfc08
R8 =ffff88000123fa60 R9 =ffff880277003a00 R10=ffffffffc01fd68c R11=0000000000000206
R12=ffff880272280880 R13=ffff880272280900 R14=ffff880272280800 R15=ffff8802752370c0
RIP=ffffffffc01fd6d5 RFL=00010246 [---Z-P-] CPL=0 II=0 A20=1 SMM=0 HLT=0
ES =0000 0000000000000000 ffffffff 00000000
CS =0010 0000000000000000 ffffffff 00a09b00 DPL=0 CS64 [-RA]
SS =0018 0000000000000000 ffffffff 00c09300 DPL=0 DS   [-WA]
DS =0000 0000000000000000 ffffffff 00000000
FS =0000 00007f42e44bc780 ffffffff 00000000
GS =0000 ffff88027fdc0000 ffffffff 00000000
LDT=0000 0000000000000000 ffffffff 00000000
TR =0040 ffff88027fdd4300 00002087 00008b00 DPL=0 TSS64-busy
GDT=     ffff88027fdc9000 0000007f
IDT=     ffffffffff574000 00000fff
CR0=80050033 CR2=0000000001681118 CR3=0000000272504000 CR4=003406e0
DR0=0000000000000000 DR1=0000000000000000 DR2=0000000000000000 DR3=0000000000000000 
DR6=00000000fffe0ff0 DR7=0000000000000400
EFER=0000000000000d01
Code=84 a9 00 00 00 31 d2 bb 03 00 00 00 b8 0a 00 00 00 48 89 f9 <0f> 01 cf 31 c0 89 45 c4 e8 3e 1e 00 00 8b 45 c4 85 c0 75 55 48 8b 45 c8 49 8d 7f 08 be 00
```

Thanks.

**Comment by andyzyb** [comment]
Can you test it in without KVM.  Looks the issue come under Guest VM?

**Comment by insujang** [comment]
Thanks for the answer. KVM internal error seems to be caused by QEMU. I re-tested from scractch, and I found out OVMF UEFI bios does not support SGX in QEMU-KVM virtual machine. KVM internal error is not printed with seabios included in QEMU-SGX.

Also, aesmd service fault is actually my fault. I am modifying Linux kernel driver source code and installing every time. Whenever I install with `sudo make modules_install` in kernel source code root directory, isgx.ko in `/lib/modules/<kernel_version>/kernel/drivers/intel/sgx` is erased. I did not notice because it works until I reboot and it shows error again after I reboot.
It seems aesmd service error is not displayed for several tests related to linux sgx driver.

So, llly was right. I unintentionally erased linux sgx driver every time after I installed it.. :/ Should have checked more carefully..!
Thanks for the help!
